### PR TITLE
Initial commit for the email crawler

### PR DESCRIPTION
**Pull request by phaedrus**
This pull implements a simple crawler that when given a domain (like mit.edu) will crawl pages on that domain and pull email addresses from them. Each email address is printed to the console.

To build the project I've provided a simple build script:

./build.sh

You can run the crawler using the run script. For example:

./run.sh mit.edu

### CODE REVIEW

File: src/main/java/com/emailcrawling/crawler/Crawler.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    this.crawlDomain = crawlDomain;
+  }
+
+  @Override
+  public boolean shouldVisit(Page referringPage, WebURL url) {
+    String lowerCaseUrl = url.getURL().toLowerCase();
+    return !isMediaUrl(lowerCaseUrl) && matchesDomain(url);
+  }
+
+  private boolean isMediaUrl(final String url) {
```
Comment (clee-jana) [inline src/main/java/com/emailcrawling/crawler/Crawler.java:45]: good catch on the media urls

File: src/main/java/com/emailcrawling/crawler/Crawler.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    return MEDIA_FILE_PATTERN.matcher(url).matches() || url.contains("/img/") || url.contains("/css/");
```
Comment (clee-jana) [inline src/main/java/com/emailcrawling/crawler/Crawler.java]: why do you think we also need to check the path? do you think just checking the media extension is enough?

File: src/main/java/com/emailcrawling/crawler/CrawlerModule.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  static PageFetcher pageFetcher(CrawlConfig config) {
+    return new PageFetcher(config);
+  }
+
+  @Provides
+  static CrawlConfig crawlConfig() {
+    String crawlStorageFolder = "/tmp/.emailcrawl/";
+    CrawlConfig config = new CrawlConfig();
+    //config.setPolitenessDelay(100);
+    config.setConnectionTimeout(3000);
```
Comment (clee-jana) [inline src/main/java/com/emailcrawling/crawler/CrawlerModule.java:42]: 👍  for setting a connection timeout

File: src/test/java/com/emailcrawling/crawler/TestCrawler.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
+package com.emailcrawling.crawler;
```
Comment (clee-jana) [inline src/test/java/com/emailcrawling/crawler/TestCrawler.java:1]: 👍  for having tests!

File: src/main/java/com/emailcrawling/app/App.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    return Flowable
+            .create(e -> {
+              startCrawl(domain);
+              e.onNext(new Object());
+              controller.waitUntilFinish();
+              pagePublisher.complete();
+              e.onComplete();
+              controller.shutdown();
+
+            }, BackpressureStrategy.MISSING)
```
Comment (clee-jana) [inline src/main/java/com/emailcrawling/app/App.java:66]: I'm not that familiar with RxJava, but what's a `BackpressureStrategy` and why did you choose `MISSING`?

**Review by clee-jana — APPROVED** [review]
overall looks really solid to me. i asked some questions below if you wouldn't mind answering them.

I noticed on my end that when I run the program, it doesn't return back to the command line, it seems to get stuck? for example when I run it on jana.com it outputs the correct emails but never exits?

File: src/main/java/com/emailcrawling/app/App.java (hunk 5 above)
Comment (phaedrus) [reply src/main/java/com/emailcrawling/app/App.java:66]: Backpressure occurs when the source observable emits more items than can be consumed downstream. In this case we only ever emit one item when we start the crawl so there shouldn't be any backpressure here. I added a back pressure buffer stream further downstream where each of the pages is emitted. I used the buffer strategy so we wouldn't lose any pages.

File: src/main/java/com/emailcrawling/crawler/Crawler.java (hunk 2 above)
Comment (phaedrus) [reply src/main/java/com/emailcrawling/crawler/Crawler.java]: Yah I thought it wasn't catching everything so I looked on github to see how other people were filtering out urls with this library and found this. Looking at it now I don't think it makes a lot of sense.

**Comment by phaedrus** [comment]
Hmm...it exited for me last night but let me try again now with logging turned up.

**Comment by phaedrus** [comment]
Ok so I re-read the requirements and I was probably crawling too much (in particular it was crawling subdomains like (blog.jana.com)). It looks like this is because I'm not doing the domain matching correctly.
What had confused me initially was that the example output for mit.edu. I had initially typed http://mit.edu and it redirected immediately to web.mit.edu which would have been forbidden by my domain match so I may have made that too flexible. 
I think the right fix would be to match the url subdomain against the www. and web. variants of the input domain. Something like this:

```
  private boolean matchesDomain(WebURL url) {
    String subDomain = (url.getSubDomain() + url.getDomain()).toLowerCase();
    return subDomain.equals(crawlDomain) || subDomain.equals(webCrawlDomain) || subDomain.equals(wwwCrawlDomain);
  }
```
I pushed an additional commit with a fix as well.
